### ISSUE TITLE
`SIM108` false negative when variable has type annotation

### ISSUE THREAD

**Issue by nickdrozd** [astral-sh/ruff#15554]
Assignment with `if/else` gets flagged with `SIM108`, suggesting ternary operator. Good. But if the variable is annotated, not flagged:

```python
from random import randint

if randint(0, 3):
    x: int = 4  # not flagged
else:
    x = 5
```

Could be `if` var annotated or `else` var, not flagged. Also not flagged if both vars have annotation, even if annotation is the same.

**Comment by MichaReiser** [comment]
That makes sense to me but is slightly more complicated because that only works if both branches infer to the same type, unless the fix removes the type annotation.

**Comment by nickdrozd** [comment]
If the variable is declared with different types in the branches, Mypy flags it as a redefinition:

```python
if randint(0, 3):
    x: int = 4
else:
    x: str = 'q'  # mypy flags as redefined
```

(It's not even clear whether there is one variable `x` defined in two places or two separate variables each called `x`. Yet another reason not to define a variable in branches.)

In this case it would be fine to add a new annotation that is the union of types in the branches:

```python
x: int | str = 4 if randint(0, 3) else 'q'
```

**Comment by MichaReiser** [comment]
The problem I see is more about:

```
if randint(0, 3):
    x: int = 4
else:
    x = 'q'  # mypy flags as redefined
```

Mypy might also flag this as redefinition but I'd have to check if this is true for every type checker (including red knot). But I could still see us flagging this but we may not want to provide a fix for it.

**Comment by JelleZijlstra** [comment]
Right now, both ty and pyrefly allow this program:

```python
from typing import reveal_type
from random import randint

if randint(0, 3):
    x: int = 4  # not flagged
else:
    x = "x"
```

So "simplifying" this is questionable.
